### PR TITLE
Convert EFA installation recipe into a resource

### PR DESCRIPTION
**Pull request by enrico-usai**
### Description of changes
- Convert EFA installation recipe into a resource
- Split EFA resource to isolate OS specific code
  * I'm using an action_class to define the OS specific conflicting_packages attribute.

### Tests
- Tested on EC2 on all OSes with a t2.medium, because t2.micro is not enough to install EFA on RHEL8
- Fixed EFA kernel module check to be compatible with CentOS7

### CODE REVIEW

File: cookbooks/aws-parallelcluster-install/resources/efa/partial/_setup.rb [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  efa_tarball = "#{node['cluster']['sources_dir']}/aws-efa-installer.tar.gz"
+  efa_installer_url = "https://efa-installer.amazonaws.com/aws-efa-installer-#{node['cluster']['efa']['installer_version']}.tar.gz"
+
+  # Get EFA Installer
+  remote_file efa_tarball do
+    source efa_installer_url
+    mode '0644'
+    retries 3
+    retry_delay 5
+    not_if { ::File.exist?(efa_tarball) }
```
Comment (awslitvit) [inline cookbooks/aws-parallelcluster-install/resources/efa/partial/_setup.rb]: Nitpick: you can use `action :create_if_missing` to this end.

File: cookbooks/aws-parallelcluster-install/libraries/helpers.rb [hunk 2]
```diff
@@ -90,3 +90,13 @@ def append_if_not_present_grub_cmdline(attributes, grub_variable)
...
+
+def efa_installed?
+  dir_exist = ::Dir.exist?('/opt/amazon/efa')
+  if dir_exist
+    modinfo_efa_stdout = Mixlib::ShellOut.new("modinfo efa").run_command.stdout
+    efa_installed_packages_file = Mixlib::ShellOut.new("cat /opt/amazon/efa_installed_packages").run_command.stdout
+    Chef::Log.info("`/opt/amazon/efa` directory already exists. \nmodinfo efa stdout: \n#{modinfo_efa_stdout} \nefa_installed_packages_file_content: \n#{efa_installed_packages_file}")
+  end
+  dir_exist
+end
```
Comment (awslitvit) [inline cookbooks/aws-parallelcluster-install/libraries/helpers.rb]: Thought: I just wonder if this shouldn't go to networking library in common cookbook.

**Review by awslitvit — DISMISSED** [review]

File: cookbooks/aws-parallelcluster-install/resources/efa/partial/_setup.rb (hunk 1 above)
Comment (enrico-usai) [reply cookbooks/aws-parallelcluster-install/resources/efa/partial/_setup.rb]: done, thanks!

File: cookbooks/aws-parallelcluster-install/libraries/helpers.rb (hunk 2 above)
Comment (enrico-usai) [reply cookbooks/aws-parallelcluster-install/libraries/helpers.rb]: Moved, good point!

File: cookbooks/aws-parallelcluster-common/libraries/helpers.rb [hunk 3]
```diff
@@ -19,3 +19,7 @@ def x86?
 def arm_instance?
   node['kernel']['machine'] == 'aarch64'
 end
+
+def efa_supported?
+  !arm_instance? || !default['cluster']['efa']['unsupported_aarch64_oses'].include?(node['cluster']['base_os'])
+end
```
Comment (awslitvit) [inline cookbooks/aws-parallelcluster-common/libraries/helpers.rb]: This one should also be moved to networking library.

**Review by awslitvit — APPROVED** [review]

File: cookbooks/aws-parallelcluster-common/libraries/helpers.rb (hunk 3 above)
Comment (enrico-usai) [reply cookbooks/aws-parallelcluster-common/libraries/helpers.rb]: done

**Pull request merged**